### PR TITLE
feat: swapper, reborn

### PR DESCRIPTION
**Pull request by 0xApotheosis**
## Description

Adds a new, rearchitected, swapper behind a feature flag `SwapperV2`.

The new swapper:

- Uses RTK Query for requests, allowing us to cache and poll in a first-class way
- Refactors swapper architecture to use micro-services where:
  - All side effects are handled by dedicated services
  - All stateful computations are handled and decoupled via custom hooks
  - All non-stateful calculations are handled by pure utility functions (_no_ side-effects or implicit arguments)

These is one visual change that is visible outside of the feature flag, the new confirmation screen:

**Before**

Screen Shot 2022-08-30 at 5 58 26 pm

**After**

Screen Shot 2022-08-30 at 5 57 58 pm

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

Contributes to https://github.com/shapeshift/web/issues/2556

## Risk

Most of this code is new and only touches the V2 swapper. That said, some refactoring was done that is consumed by V1 swapper, so we need to confirm there are no regressions before merging this.

### Engineering

This is an incremental PR to keep the diff small for reviewers (it's already massive, sorry). 
It is not intended to be fully functional, but it is intended to meet best practice code and architecture standards. Please review accordingly.

Most important is no regressions against swapper V1.

### Operations

- No regressions should occur for the existing V1 swapper
- New "confirm trade" screen (shown after clicking "Preview Trade") functions as expected, showing sane values in all fields, including the new "Expected Amount" dropdown

## Screenshots (if applicable)

Screen Shot 2022-08-29 at 2 48 50 pm

### CODE REVIEW

File: src/components/Trade/services/useTradeQuoteService.ts [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+  isSupportedNoneUtxoSwappingChain,
+  isSupportedUtxoSwappingChain,
+  TradeQuoteInputCommonArgs,
+} from 'components/Trade/hooks/useSwapper/utils'
+import { TradeState } from 'components/Trade/types'
+import { getChainAdapterManager } from 'context/PluginProvider/chainAdapterSingleton'
+import { useWallet } from 'hooks/useWallet/useWallet'
+import { toBaseUnit } from 'lib/math'
+import { useGetTradeQuoteQuery } from 'state/apis/swapper/swapperApi'
+import { selectAccountSpecifiers } from 'state/slices/accountSpecifiersSlice/selectors'
```
Comment (0xdef1cafe) [inline src/components/Trade/services/useTradeQuoteService.ts:23]: let's move away from this selector and query this from the portfolioSlice so we can kill the accountSpecifiersSlice entirely

File: src/components/Trade/hooks/useSwapper/utils.ts [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+  }
+}
+
+export const getBestSwapperFromArgs = async (
+  buyAssetId: AssetId,
+  sellAssetId: AssetId,
+  featureFlags: FeatureFlags,
+): Promise<Swapper<ChainId>> => {
+  const swapperManager = await getSwapperManager(featureFlags)
+  const swapper = await swapperManager.getBestSwapper({
```
Comment (0xdef1cafe) [inline src/components/Trade/hooks/useSwapper/utils.ts:203]: just be aware the current logic of "best swapper" is the 0th in the list

File: src/components/Trade/services/useTradeQuoteService.ts (hunk 1 above)
Comment (0xApotheosis) [reply src/components/Trade/services/useTradeQuoteService.ts]: Roger that. I'll sort it out in another PR soon.

File: src/components/Trade/hooks/useSwapper/utils.ts (hunk 2 above)
Comment (0xApotheosis) [reply src/components/Trade/hooks/useSwapper/utils.ts:197]: Yes indeed. Not for this PR, but it'll get looked at 👌

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
Added smallish, non-blocking comments.

Didn't focus on testing the new swapper feature under flag, as the goal is to get the core work in develop as a base while ensuring no regressions, and it isn't fully functional yet (wasn't able to get into an input state from dashboard, and when inputting from an asset page, I wasn't able to get past the input route).

Regression tested:

- Swapper input for all the following trades
- Swapper quote/preview for all the following trades
- Swapper trade pending/confirmation state for all the following trades

- AVAX -> WETH on Avalanche (0x, MetaMask)
- WETH on Avalanche -> AVAX  (0x, MetaMask)
- ETH -> FOX (0x, MetaMask)
- FOX -> WETH (CowSwap, MetaMask)
- WETH -> ETH (CowSwap, MetaMask)
- ATOM -> OSMO (Thor, Native)
- OSMO -> ATOM (Thor, Native)

Spotted regressions:

- FOX is always hardcoded as "Expected Amount" on trade preview and confirmation routes

image
image
image
image

File: src/components/Trade/TradeConfirm/TradeConfirm.tsx [hunk 3]
```diff
@@ -168,109 +194,112 @@ export const TradeConfirm = ({ history }: RouterProps) => {
...
+                buyIcon={trade.buyAsset.icon}
+                sellIcon={trade.sellAsset.icon}
+                status={sellTxid || isSubmitting ? status : undefined}
+              />
+              <Stack spacing={4}>
                 <Row>
-                  <Row.Label>
-                    <RawText>Tx ID</RawText>
-                  </Row.Label>
+                  <Row.Label>Send</Row.Label>
```
Comment (0xdef1cafe) [inline src/components/Trade/TradeConfirm/TradeConfirm.tsx:214]: translation

File: src/components/DeFi/components/AssetInput.tsx [hunk 4]
```diff
@@ -139,7 +141,13 @@ export const AssetInput: React.FC<AssetInputProps> = ({
 
       {showFiatAmount && (
         <Stack width='full' alignItems='flex-end' px={4} pb={2}>
-          <Button onClick={() => setIsFiat(!isFiat)} size='xs' variant='link' colorScheme='blue'>
+          <Button
+            onClick={() => setIsFiat(!isFiat)}
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/components/DeFi/components/AssetInput.tsx:145]: Unless we have something like the below, it's not a toggle:

```ts
const toggleIsFiat = () => {
  setIsFiat(current => !current);
}
```

File: src/components/Trade/TradeInputV2.tsx [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+              if (isFiat) {
+                fiatBuyAmount.onChange(value)
+              } else {
+                cryptoBuyAmount.onChange(value)
+              }
+              handleInputChange(action, value)
+            }}
+            percentOptions={[1]}
+            onAssetClick={() => history.push(TradeRoutePaths.BuySelect)}
+          ></TradeAssetInput>
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/components/Trade/TradeInputV2.tsx]: Fixed in https://github.com/shapeshift/web/pull/2584/commits/54b83b6222ece5bcc87a545720cd3fae61ec0dfc

I wonder if we can lint for this.

File: src/components/DeFi/components/AssetInput.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/components/DeFi/components/AssetInput.tsx]: Added `useToggle` in https://github.com/shapeshift/web/pull/2584/commits/54b83b6222ece5bcc87a545720cd3fae61ec0dfc - I like it!

[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/components/DeFi/components/AssetInput.tsx]: Yeah this is the go, added in https://github.com/shapeshift/web/pull/2584/commits/54b83b6222ece5bcc87a545720cd3fae61ec0dfc

File: src/theme/theme.tsx [hunk 6]
```diff
@@ -39,6 +40,7 @@ const styles = {
     body: {
       backgroundColor: mode('gray.50', 'gray.800')(props),
       backgroundSize: 'cover',
+      'font-feature-settings': "'zero'",
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/theme/theme.tsx]: Good catch!

Addressed in https://github.com/shapeshift/web/pull/2584/commits/c5109ec193eb7bec5bc2c42256a348b52318b052

File: src/components/Trade/TradeInputV2.tsx [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+    selectFeeAssetById(state, sellTradeAsset?.asset?.assetId ?? ethAssetId),
+  )
+  const sellAssetBalance = useAppSelector(state =>
+    selectPortfolioCryptoBalanceByFilter(state, {
+      accountId: sellAssetAccount,
+      assetId: sellTradeAsset?.asset?.assetId ?? '',
+    }),
+  )
+
+  // const { updateTrade, checkApprovalNeeded } = useSwapper()
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/components/Trade/TradeInputV2.tsx]: Addressed in https://github.com/shapeshift/web/pull/2584/commits/c5109ec193eb7bec5bc2c42256a348b52318b052

File: src/components/Trade/Components/TradeAssetInput.tsx [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+  const fiatBalance = bnOrZero(balance).times(marketData.price).toString()
+
+  return <AssetInput balance={balance} fiatBalance={fiatBalance} {...rest} />
+}
+
+type TradeAssetInputProps = {
+  assetId?: AssetId
+} & AssetInputProps
+
+export const TradeAssetInput: React.FC<TradeAssetInputProps> = props => {
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/components/Trade/Components/TradeAssetInput.tsx]: That's a much nicer pattern 👌 

Addressed in https://github.com/shapeshift/web/pull/2584/commits/c5109ec193eb7bec5bc2c42256a348b52318b052

File: src/components/Trade/Components/RateGasRow.tsx [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+  ) : (
+    <Stack direction='row' fontWeight='medium'>
+      <Row fontSize='sm' flex={1}>
+        <Row.Value fontSize='sm'>
+          <HelperTooltip
+            label={translate('trade.tooltip.rate')}
+            flexProps={{ flexDirection: 'row-reverse' }}
+          >
+            <Stack width='full' direction='row' spacing={1}>
+              <Amount.Crypto fontSize='sm' value='1' symbol={sellSymbol ?? ''} suffix='=' />
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/components/Trade/Components/RateGasRow.tsx]: It should!

Addressed in https://github.com/shapeshift/web/pull/2584/commits/c5109ec193eb7bec5bc2c42256a348b52318b052

File: src/components/Trade/TradeConfirm/TradeConfirm.tsx (hunk 3 above)
Comment (0xApotheosis) [reply src/components/Trade/TradeConfirm/TradeConfirm.tsx]: Addressed in https://github.com/shapeshift/web/pull/2584/commits/b1d5491d937fc68569f7781d2b1e93d27c47575c

File: src/components/Trade/TradeInputV2.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply src/components/Trade/TradeInputV2.tsx]: You bet https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/self-closing-comp.md

File: src/components/Trade/TradeInputV2.tsx [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+    setValue('amount', maxSendAmount)
+  }
+
+  const onSubmit = async (values: TradeState<KnownChainIds>) => {
+    console.info(values)
+    try {
+      // TODO: Check if approval needed
+      history.push({ pathname: TradeRoutePaths.Confirm, state: { fiatRate: feeAssetFiatRate } })
+    } catch (e) {
+      console.error(e)
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/components/Trade/TradeInputV2.tsx]: https://github.com/shapeshift/web/pull/2584/commits/92f1b2701af78fd37a565c7ae7b7b3654fa4869b

File: src/components/Trade/TradeInputV2.tsx [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+      quote,
+      sellAssetBalance,
+    )
+    setValue('action', TradeAmountInputField.SELL_CRYPTO)
+    setValue('sellTradeAsset.amount', maxSendAmount)
+    setValue('amount', maxSendAmount)
+  }
+
+  const onSubmit = async (values: TradeState<KnownChainIds>) => {
+    console.info(values)
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/components/Trade/TradeInputV2.tsx]: https://github.com/shapeshift/web/pull/2584/commits/92f1b2701af78fd37a565c7ae7b7b3654fa4869b

File: src/components/Trade/Components/RateGasRow.tsx [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+type RateGasRowProps = {
+  sellSymbol?: string
+  buySymbol?: string
+  rate?: string
+  gasFee: string
+}
+export const RateGasRow: FC<RateGasRowProps> = ({ sellSymbol, buySymbol, rate, gasFee }) => {
+  const translate = useTranslate()
+  return !rate ? (
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/components/Trade/Components/RateGasRow.tsx:21]: I'm down to bikeshed this one day 😁 

I haven't yet come around to using multiple return points where a ternary will do, as it breaks the functional paradigm.

**Pull request merged**